src/main/java/org/terasology/moduletestingenvironment/ModuleTestingHelper.java

```diff
@@ -1,15 +1,158 @@
-// Copyright 2020 The Terasology Foundation
+// Copyright 2021 The Terasology Foundation
 // SPDX-License-Identifier: Apache-2.0
-
 package org.terasology.moduletestingenvironment;
 
+import com.google.common.util.concurrent.ListenableFuture;
+import org.joml.Vector3fc;
+import org.joml.Vector3ic;
+import org.terasology.engine.context.Context;
+import org.terasology.engine.core.TerasologyEngine;
+import org.terasology.engine.core.modes.StateIngame;
+import org.terasology.engine.world.block.BlockRegion;
+import org.terasology.engine.world.block.BlockRegionc;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.function.Supplier;
+
 /**
- * Helpers for use in module tests.
+ * Base class for tests involving full {@link TerasologyEngine} instances. View the tests included in this module for
+ * simple usage examples.
+ *
+ * <h2>Introduction</h2>
+ * If test classes extend this class will create a new host engine for each {@code @Test} method. If the testing
+ * environment is used by composition {@link #setup()} and {@link #tearDown()} need to be called explicitly. This can be
+ * done once for the test class or for each test.
  * <p>
- * See {@link ModuleTestingEnvironment} for available methods
+ * The in-game {@link Context} for this engine can be accessed via {@link #getHostContext()}. The result of this getter
+ * is equivalent to the CoreRegistry available to module code at runtime. However, it is very important that you do not
+ * use CoreRegistry in your test code, as this is manipulated by the test environment to allow multiple instances of the
+ * engine to peacefully coexist. You should always use the returned context reference to manipulate or inspect the
+ * CoreRegistry of a given engine instance.
+ *
+ * <h2>Client Engine Instances</h2>
+ * Client instances can be easily created via {@link #createClient()} which returns the in-game context of the created
+ * engine instance. When this method returns, the client will be in the {@link StateIngame} state and connected to the
+ * host. Currently all engine instances are headless, though it is possible to use headed engines in the future.
  * <p>
- * This dummy subclass exists to provide a non-deprecated interface to the helper functions in the base class.
- * In the next major release the base class will be removed and all methods will live in this class directly.
+ * Engines can be run while a condition is true via {@link #runWhile(Supplier)} <br>{@code runWhile(()-> true);}
+ * <p>
+ * or conversely run until a condition is true via {@link #runUntil(Supplier)} <br>{@code runUntil(()-> false);}
+ *
+ * <h2>Specifying Dependencies</h2>
+ * By default the environment will load only the engine itself. FIXME
```

Comment:
Maybe these two FIXMEs can be moved around a bit so there's actual descriptive text behind them?

Reply:
I've pushed a big update to the javadoc.